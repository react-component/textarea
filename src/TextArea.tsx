import clsx from 'classnames';
import { BaseInput } from 'rc-input';
import {
  fixControlledValue,
  resolveOnChange,
} from 'rc-input/lib/utils/commonUtils';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import type { ReactNode } from 'react';
import React, { useEffect, useImperativeHandle, useRef } from 'react';
import type {
  ResizableTextAreaRef,
  TextAreaProps,
  TextAreaRef,
} from './interface';
import ResizableTextArea from './ResizableTextArea';

function fixEmojiLength(value: string, maxLength: number) {
  return [...(value || '')].slice(0, maxLength).join('');
}

function setTriggerValue(
  isCursorInEnd: boolean,
  preValue: string,
  triggerValue: string,
  maxLength: number,
) {
  let newTriggerValue = triggerValue;
  if (isCursorInEnd) {
    // 光标在尾部，直接截断
    newTriggerValue = fixEmojiLength(triggerValue, maxLength!);
  } else if (
    [...(preValue || '')].length < triggerValue.length &&
    [...(triggerValue || '')].length > maxLength!
  ) {
    // 光标在中间，如果最后的值超过最大值，则采用原先的值
    newTriggerValue = preValue;
  }
  return newTriggerValue;
}

const TextArea = React.forwardRef<TextAreaRef, TextAreaProps>(
  (
    {
      defaultValue,
      value: customValue,
      onFocus,
      onBlur,
      onChange,
      allowClear,
      maxLength,
      onCompositionStart,
      onCompositionEnd,
      suffix,
      prefixCls = 'rc-textarea',
      classes,
      showCount,
      className,
      style,
      disabled,
      hidden,
      classNames,
      styles,
      onResize,
      ...rest
    },
    ref,
  ) => {
    const [value, setValue] = useMergedState(defaultValue, {
      value: customValue,
      defaultValue,
    });
    const resizableTextAreaRef = useRef<ResizableTextAreaRef>(null);

    const [focused, setFocused] = React.useState<boolean>(false);

    const [compositing, setCompositing] = React.useState(false);
    const oldCompositionValueRef = React.useRef<string>();
    const oldSelectionStartRef = React.useRef<number>(0);
    // Since ResizeObserver would resize once on mounted, manual resizing should be happened after that
    const [resizeStatus, setResizeStatus] = React.useState<
      'mounted' | 'resized' | null
    >(null);

    const focus = () => {
      resizableTextAreaRef.current.textArea.focus();
    };

    useImperativeHandle(ref, () => ({
      resizableTextArea: resizableTextAreaRef.current,
      focus,
      blur: () => {
        resizableTextAreaRef.current.textArea.blur();
      },
    }));

    useEffect(() => {
      setFocused((prev) => !disabled && prev);
    }, [disabled]);

    // =========================== Value Update ===========================
    // Max length value
    const hasMaxLength = Number(maxLength) > 0;

    const onInternalCompositionStart: React.CompositionEventHandler<
      HTMLTextAreaElement
    > = (e) => {
      setCompositing(true);
      // 拼音输入前保存一份旧值
      oldCompositionValueRef.current = value as string;
      // 保存旧的光标位置
      oldSelectionStartRef.current = e.currentTarget.selectionStart;
      onCompositionStart?.(e);
    };

    const onInternalCompositionEnd: React.CompositionEventHandler<
      HTMLTextAreaElement
    > = (e) => {
      setCompositing(false);

      let triggerValue = e.currentTarget.value;
      if (hasMaxLength) {
        const isCursorInEnd =
          oldSelectionStartRef.current >= maxLength! + 1 ||
          oldSelectionStartRef.current ===
            oldCompositionValueRef.current?.length;
        triggerValue = setTriggerValue(
          isCursorInEnd,
          oldCompositionValueRef.current as string,
          triggerValue,
          maxLength!,
        );
      }
      // Patch composition onChange when value changed
      if (triggerValue !== value) {
        setValue(triggerValue);
        resolveOnChange(e.currentTarget, e, onChange, triggerValue);
      }

      onCompositionEnd?.(e);
    };

    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
      let triggerValue = e.target.value;
      if (!compositing && hasMaxLength) {
        // 1. 复制粘贴超过maxlength的情况 2.未超过maxlength的情况
        const isCursorInEnd =
          e.target.selectionStart >= maxLength! + 1 ||
          e.target.selectionStart === triggerValue.length ||
          !e.target.selectionStart;
        triggerValue = setTriggerValue(
          isCursorInEnd,
          value as string,
          triggerValue,
          maxLength!,
        );
      }
      setValue(triggerValue);
      resolveOnChange(e.currentTarget, e, onChange, triggerValue);
    };

    const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
      const { onPressEnter, onKeyDown } = rest;
      if (e.key === 'Enter' && onPressEnter) {
        onPressEnter(e);
      }
      onKeyDown?.(e);
    };

    const handleFocus: React.FocusEventHandler<HTMLTextAreaElement> = (e) => {
      setFocused(true);
      onFocus?.(e);
    };

    const handleBlur: React.FocusEventHandler<HTMLTextAreaElement> = (e) => {
      setFocused(false);
      onBlur?.(e);
    };

    // ============================== Reset ===============================
    const handleReset = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
      setValue('');
      focus();
      resolveOnChange(resizableTextAreaRef.current.textArea, e, onChange);
    };

    let val = fixControlledValue(value);

    if (
      !compositing &&
      hasMaxLength &&
      (customValue === null || customValue === undefined)
    ) {
      // fix #27612 将value转为数组进行截取，解决 '😂'.length === 2 等emoji表情导致的截取乱码的问题
      val = fixEmojiLength(val, maxLength!);
    }

    let suffixNode = suffix;
    let dataCount: ReactNode;
    if (showCount) {
      const valueLength = [...val].length;

      if (typeof showCount === 'object') {
        dataCount = showCount.formatter({
          value: val,
          count: valueLength,
          maxLength,
        });
      } else {
        dataCount = `${valueLength}${hasMaxLength ? ` / ${maxLength}` : ''}`;
      }

      suffixNode = (
        <>
          {suffixNode}
          <span
            className={clsx(`${prefixCls}-data-count`, classNames?.count)}
            style={styles?.count}
          >
            {dataCount}
          </span>
        </>
      );
    }

    const handleResize: TextAreaProps['onResize'] = (size) => {
      onResize?.(size);
      if (resizeStatus === null) {
        setResizeStatus('mounted');
      } else if (resizeStatus === 'mounted') {
        setResizeStatus('resized');
      }
    };

    const textarea = (
      <BaseInput
        value={val}
        allowClear={allowClear}
        handleReset={handleReset}
        suffix={suffixNode}
        prefixCls={prefixCls}
        classes={{
          affixWrapper: clsx(classes?.affixWrapper, {
            [`${prefixCls}-show-count`]: showCount,
            [`${prefixCls}-textarea-allow-clear`]: allowClear,
          }),
        }}
        disabled={disabled}
        focused={focused}
        className={className}
        style={{
          ...style,
          ...(resizeStatus === 'resized' ? { height: 'auto' } : {}),
        }}
        dataAttrs={{
          affixWrapper: {
            'data-count': typeof dataCount === 'string' ? dataCount : undefined,
          },
        }}
        hidden={hidden}
        inputElement={
          <ResizableTextArea
            {...rest}
            onKeyDown={handleKeyDown}
            onChange={handleChange}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onCompositionStart={onInternalCompositionStart}
            onCompositionEnd={onInternalCompositionEnd}
            className={classNames?.textarea}
            style={{ ...styles?.textarea, resize: style?.resize }}
            disabled={disabled}
            prefixCls={prefixCls}
            onResize={handleResize}
            ref={resizableTextAreaRef}
          />
        }
      />
    );

    return textarea;
  },
);

export default TextArea;
