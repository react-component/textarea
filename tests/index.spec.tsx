import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import type { TextAreaProps } from '../src';
import TextArea from '../src';
import calculateAutoSizeStyle, {
  calculateNodeStyling,
} from '../src/calculateNodeHeight';
import { triggerResize, wait } from './utils';

describe('TextArea', () => {
  const originalGetComputedStyle = window.getComputedStyle;
  beforeAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: (node) => ({
        getPropertyValue: (prop) => {
          if (prop === 'box-sizing') {
            return originalGetComputedStyle(node)[prop] || 'border-box';
          }
          return originalGetComputedStyle(node)[prop];
        },
      }),
    });
  });

  afterAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: originalGetComputedStyle,
    });
  });

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllTimers();
    jest.useRealTimers();
  });

  it('should work correctly on controlled mode', () => {
    const Demo = () => {
      const [value, setValue] = React.useState('111');
      return (
        <TextArea
          value={value}
          onChange={(e) => {
            setValue(e.target.value);
          }}
        />
      );
    };
    const { container } = render(<Demo />);
    const textarea = container.querySelector('textarea');
    expect(textarea.value).toBe('111');
    fireEvent.change(textarea, {
      target: {
        value: '222',
      },
    });
    expect(textarea.value).toBe('222');
  });

  it('should auto calculate height according to content length and autoSize property', async () => {
    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const onInternalAutoSize = jest.fn();

    const renderDemo = (
      value = '',
      autoSize: TextAreaProps['autoSize'] = { minRows: 2, maxRows: 6 },
    ) => (
      <TextArea
        value={value}
        readOnly
        autoSize={autoSize}
        wrap="off"
        // @ts-ignore
        onInternalAutoSize={onInternalAutoSize}
      />
    );

    const { container, rerender } = render(renderDemo());

    await wait();
    onInternalAutoSize.mockRestore();

    rerender(renderDemo('1111\n2222\n3333'));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(1);

    rerender(renderDemo('1111'));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(2);

    rerender(renderDemo('1111', { minRows: 1, maxRows: 6 }));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(3);

    rerender(renderDemo('1111', { minRows: 1, maxRows: 6 }));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(3);

    rerender(renderDemo('1111', { minRows: 1, maxRows: 12 }));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(4);

    rerender(renderDemo('1111', true));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(5);

    rerender(renderDemo('1111', false));
    await wait();
    expect(onInternalAutoSize).toHaveBeenCalledTimes(5);

    expect(container.querySelector('textarea')).not.toHaveStyle({
      overflowY: 'hidden',
    });

    expect(errorSpy).not.toHaveBeenCalled();
  });

  it('should support onPressEnter and onKeyDown', () => {
    const fakeHandleKeyDown = jest.fn();
    const fakeHandlePressEnter = jest.fn();
    const { container } = render(
      <TextArea
        onKeyDown={fakeHandleKeyDown}
        onPressEnter={fakeHandlePressEnter}
      />,
    );
    const textarea = container.querySelector('textarea');
    /** keyCode 65 is A */
    fireEvent.keyDown(textarea, { keyCode: 65 });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(1);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(0);

    /** keyCode 13 is Enter */
    fireEvent.keyDown(textarea, { keyCode: 13 });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(2);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(1);
  });

  it('should support disabled', () => {
    const { container } = render(<TextArea disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should support maxLength', () => {
    const { container } = render(<TextArea maxLength={10} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calculateNodeStyling works correctly', () => {
    const wrapper = document.createElement('textarea');
    wrapper.id = 'test';
    wrapper.wrap = 'wrap';
    calculateNodeStyling(wrapper, true);
    const value = calculateNodeStyling(wrapper, true);
    expect(value).toEqual({
      borderSize: 2,
      boxSizing: 'border-box',
      paddingSize: 4,
      sizingStyle:
        'letter-spacing:normal;line-height:normal;padding-top:2px;padding-bottom:2px;font-family:-webkit-small-control;font-weight:;font-size:;font-variant:;text-rendering:auto;text-transform:none;width:;text-indent:0;padding-left:2px;padding-right:2px;border-width:1px;box-sizing:border-box;word-break:;white-space:pre-wrap',
    });
  });

  it('boxSizing === "border-box"', () => {
    const wrapper = document.createElement('textarea');
    wrapper.style.boxSizing = 'border-box';
    const { height } = calculateAutoSizeStyle(wrapper);
    expect(height).toBe(2);
  });

  it('boxSizing === "content-box"', () => {
    const wrapper = document.createElement('textarea');
    wrapper.style.boxSizing = 'content-box';
    const { height } = calculateAutoSizeStyle(wrapper);
    expect(height).toBe(-4);
  });

  it('minRows or maxRows is not null', () => {
    const wrapper = document.createElement('textarea');
    expect(calculateAutoSizeStyle(wrapper, false, 1, 1)).toEqual(
      expect.objectContaining({
        height: 2,
        minHeight: 2,
        overflowY: 'hidden',
        resize: 'none',
      }),
    );
    wrapper.style.boxSizing = 'content-box';
    expect(calculateAutoSizeStyle(wrapper, false, 1, 1)).toEqual(
      expect.objectContaining({
        height: -4,
        minHeight: -4,
        overflowY: 'hidden',
        resize: 'none',
      }),
    );
  });

  it('handleKeyDown', () => {
    const onPressEnter = jest.fn();
    const onKeyDown = jest.fn();
    const { container } = render(
      <TextArea
        onPressEnter={onPressEnter}
        onKeyDown={onKeyDown}
        aria-label="textarea"
      />,
    );
    fireEvent.keyDown(container.querySelector('textarea'), { keyCode: 13 });
    expect(onPressEnter).toHaveBeenCalled();
    expect(onKeyDown).toHaveBeenCalled();
  });

  it('should trigger onResize', async () => {
    const onResize = jest.fn();
    const { container } = render(<TextArea onResize={onResize} autoSize />);
    await wait();

    triggerResize(container.querySelector('textarea'));
    await wait();

    expect(onResize).toHaveBeenCalledWith(
      expect.objectContaining({
        width: expect.any(Number),
        height: expect.any(Number),
      }),
    );
  });

  it('scroll to bottom when autoSize', async () => {
    const { container, unmount } = render(<TextArea autoSize />);
    container.querySelector('textarea').focus();
    expect(document.activeElement).toBe(container.querySelector('textarea'));

    const setSelectionRangeFn = jest.spyOn(
      container.querySelector('textarea'),
      'setSelectionRange',
    );

    fireEvent.change(container.querySelector('textarea'), {
      target: { value: '\n1' },
    });

    await wait();

    expect(setSelectionRangeFn).toHaveBeenCalled();
    unmount();
  });

  it('should support onPressEnter and onKeyDown', () => {
    const fakeHandleKeyDown = jest.fn();
    const fakeHandlePressEnter = jest.fn();
    const { container } = render(
      <TextArea
        onKeyDown={fakeHandleKeyDown}
        onPressEnter={fakeHandlePressEnter}
      />,
    );
    /** KeyCode 65 is A */
    fireEvent.keyDown(container.querySelector('textarea')!, { key: 'A' });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(1);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(0);

    /** KeyCode 13 is Enter */
    fireEvent.keyDown(container.querySelector('textarea')!, { key: 'Enter' });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(2);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(1);
  });

  // 字符输入
  it('should not cut off string when cursor position is not at the end', () => {
    const onChange = jest.fn();
    const { container } = render(
      <TextArea maxLength={6} defaultValue="123456" onChange={onChange} />,
    );
    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 1, value: 'w123456' },
    });
    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 3, value: 'w123456' },
    });
    expect(container.querySelector('textarea')?.value).toBe('123456');
  });

  // 拼音输入
  // 1. 光标位于最后，且当前字符数未达到6个，若选中的字符 + 原字符的长度超过6个，则将最终的字符按照maxlength截断
  it('when the input method is pinyin and the cursor is at the end, should use maxLength to crop', () => {
    const onChange = jest.fn();
    const { container } = render(
      <TextArea maxLength={6} defaultValue="1234" onChange={onChange} />,
    );
    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 4, value: '1234' },
    });
    fireEvent.compositionStart(container.querySelector('textarea')!);

    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 9, value: '1234z z z' },
    });
    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 7, value: '1234组织者' },
    });

    fireEvent.compositionEnd(container.querySelector('textarea')!);

    expect(container.querySelector('textarea')?.value).toBe('1234组织');
  });

  // 2. 光标位于中间或开头，且当前字符数未达到6个，若选中的字符 + 原字符的长度超过6个，则显示原有字符
  it('when the input method is Pinyin and the cursor is in the middle, should display the original string', () => {
    const onChange = jest.fn();
    const { container } = render(
      <TextArea maxLength={6} defaultValue="1234" onChange={onChange} />,
    );
    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 2, value: '1234' },
    });
    fireEvent.compositionStart(container.querySelector('textarea')!);

    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 2, value: '12z z z34' },
    });
    fireEvent.change(container.querySelector('textarea')!, {
      target: { selectionStart: 5, value: '12组织者34' },
    });

    fireEvent.compositionEnd(container.querySelector('textarea')!);

    expect(container.querySelector('textarea')?.value).toBe('1234');
  });

  it('classNames and styles should work', () => {
    const { container } = render(
      <>
        <TextArea
          className="custom-class"
          style={{ background: 'red' }}
          classNames={{
            textarea: 'custom-textarea',
            count: 'custom-count',
          }}
          styles={{
            textarea: {
              color: 'red',
            },
            count: {
              color: 'blue',
            },
          }}
        />
        <TextArea
          showCount
          className="custom-class"
          style={{ background: 'red' }}
          classNames={{
            textarea: 'custom-textarea',
            count: 'custom-count',
          }}
          styles={{
            textarea: {
              color: 'red',
            },
            count: {
              color: 'blue',
            },
          }}
        />
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should be resized with textarea', async () => {
    const onResize = jest.fn();
    const { container } = render(
      <TextArea onResize={onResize} showCount style={{ height: 200 }} />,
    );
    await wait();
    expect(
      (container.querySelector('.rc-textarea-affix-wrapper') as HTMLDivElement)
        .style.height,
    ).toBe('200px');

    triggerResize(container.querySelector('textarea'), { height: 1000 });
    await wait();
    expect(onResize).toHaveBeenCalledTimes(1);
    expect(
      (container.querySelector('.rc-textarea-affix-wrapper') as HTMLDivElement)
        .style.height,
    ).toBe('auto');
  });

  it('pass style height should work', async () => {
    const onResize = jest.fn();

    const { container } = render(
      <TextArea onResize={onResize} style={{ height: 93 }} />,
    );

    triggerResize(container.querySelector('textarea'));
    await wait();

    expect(onResize).toHaveBeenCalled();

    expect(container.querySelector('textarea')).toHaveStyle({
      height: '903px',
    });
  });

  it('support bigint type', () => {
    const { container } = render(<TextArea value={BigInt('903')} autoSize />);
    expect(container.querySelector('textarea').value).toBe('903');
  });
});
