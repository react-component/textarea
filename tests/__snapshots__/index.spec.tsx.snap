// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea classNames and styles should work 1`] = `
<div>
  <textarea
    class="rc-textarea custom-textarea custom-class"
    style="color: red; background: red;"
  />
  <span
    class="rc-textarea-affix-wrapper rc-textarea-show-count custom-class"
    data-count="0"
    style="background: red;"
  >
    <textarea
      class="rc-textarea custom-textarea"
      style="color: red;"
    />
    <span
      class="rc-textarea-suffix"
    >
      <span
        class="rc-textarea-data-count custom-count"
        style="color: blue;"
      >
        0
      </span>
    </span>
  </span>
</div>
`;

exports[`TextArea should support disabled 1`] = `
<textarea
  class="rc-textarea rc-textarea-disabled"
  disabled=""
/>
`;

exports[`TextArea should support maxLength 1`] = `
<textarea
  class="rc-textarea"
  maxlength="10"
/>
`;
